### DISCUSSION TITLE
Dependency injection error

### DISCUSSION THREAD

**Discussion by herremig** [Tyrrrz/CliFx#109] [General]
Hi CliFx team,  

when I am trying to use dependency injection I get an error.  See my code below:

```csharp
   var services = new ServiceCollection();
    services.AddTransient<IExtractController, MockExtractController>();
    var serviceProvider = services.BuildServiceProvider();

    return await new CliApplicationBuilder()
        .AddCommandsFromThisAssembly()
        .UseTypeActivator(serviceProvider.GetService)
        .Build()
        .RunAsync();
```

but I keep getting the following error:

Failed to create an instance of type `Console01.ImportCommand`.
Default type activator is only capable of instantiating a type if it has a public parameterless constructor.
To fix this, either add a parameterless constructor to the type or configure a custom activator on the application.

My ImportCommand looks like the following:

```csharp
[Command]
    public class ImportCommand : ICommand
    {
        private IExtractController _controller;

        public ImportCommand(IExtractController controller)
        {
            _controller= controller;
        }

        public ValueTask ExecuteAsync(IConsole console)
        {
           
            return default;
        }
    }
```

**Comment by herremig** [comment]
fixed this, it was an easy one.  Basically I forgot to register the ImportCommand.

// Register commands
services.AddTransient<ImportCommand>();

**Reply by Tyrrrz** [reply]
Nice that you fixed it, but the error you had shouldn't happen with custom activator, it should only happen when using the default activator. Are you sure that when that error happened you had `.UseTypeActivator(serviceProvider.GetService)` configured?

**Reply by svrooij** [reply]
I can confirm the exact same thing. It will just throw an error if not all commands are registered.

We solved it by dynamicly registring all commands by looking for the already used `CommandAttribute`, not sure how fast this is though. Maybe you're better of registring all the commands statically.

```csharp
var types = Assembly.GetExecutingAssembly().GetTypes().Where(t => t.IsDefined(typeof(CliFx.Attributes.CommandAttribute)));
        foreach(var t in types) {
            services.AddTransient(t);
        }
```
